watertap/ui/api.py

```diff
@@ -493,152 +800,175 @@ def get_action(self, name):
-        # while also updating the 'missing' and 'extra' lists.
-        for data_var in variables:
-            name = data_var[cls.NAME_KEY]
-            variable_obj = getattr(ui.block, name)
-            if variable_obj is None:
-                result["missing"].append(data_var)
-            else:
-                data_val = data_var.get(cls.VALU_KEY, None)
-                if data_val is not None:
-                    if isinstance(data_val, list):
-                        for item in data_val:
-                            idx, val = tuple(item[cls.INDX_KEY]), item[cls.VALU_KEY]
-                            variable_obj[idx] = val
-                    else:
-                        variable_obj.set_value(data_val)
-                result["extra"].remove(name)
-        # return 'missing' and 'extra'
-        result["extra"] = list(result["extra"])  # normalize to lists for both
-        return result
+def _load_config(c) -> Dict:
+    if isinstance(c, dict):
+        data = c
+    else:
+        stream = open_file_or_stream(c, mode="r", encoding="utf-8")
+        data = yaml.safe_load(stream)
+    return data
+
+
+def _get_interfaces(package_name) -> Generator[Tuple[str, Callable], None, None]:
+    # import package
+    pkg = importlib.import_module(package_name)
+    # use special _ROOT variable if it exists, else package path
+    pkg_path = getattr(pkg, "_ROOT", Path(pkg.__file__).parent)
+    # if package is not a directory (e.g. a zip file), give up
+    if not pkg_path.is_dir():
+        raise IOError(f"Cannot load from package: not a directory. name={package_name}")
+    # find all python files under the package path
+    _log.info(f"Find all Python files. package-path={pkg_path}")
+    for mod_path in pkg_path.glob("**/*.py"):
+        # skip test dirs
+        if "tests" in mod_path.parts:
+            continue
+        # get name
+        nm = mod_path.name
+        # skip test files, __init__, and "protected" modules
+        if nm.startswith("_") or nm.startswith("test_"):
+            continue
+        # compute module path
+        mod_relpath = mod_path.relative_to(pkg_path)
+        # convert from posix path to module name by stripping ".py"
+        # suffix and replacing slashes with dots
+        mod_relname = mod_relpath.as_posix()[:-3].replace("/", ".")
+        mod_name = package_name + "." + mod_relname
+        # import the module
+        _log.debug(f"Importing module. name={mod_name}")
+        try:
+            mod = importlib.import_module(mod_name)
+        except Exception as err:
+            _log.warning(f"Skipping module due to import error: {err}. name={mod_name}")
+            continue
+        # look for special interface function
+        func = getattr(mod, "flowsheet_interface", None)
+        # if found, yield this module and function as a result
+        if func is not None:
+            yield mod_name, func
```

Comment:
So, if I am following this correctly, then to make our flowsheets/modules discoverable in this function we just need to provide the appropriate `flowsheet_interface` method, correct?

Should we start creating these methods in our unit models/modules by default? Should we incorporate this into our coding standards?

Reply:
Yes, that's the idea, but in a bit. There will be a UI demo at the end of the month. In doing that I expect to come across more mistakes and rough edges, and then in July we should be ready to push this a little more broadly. Though I still would think of this as "beta" since we need to see how it translates to usage in Jupyter notebooks etc.